### PR TITLE
refactor: improve consistent access under "remove if"

### PR DESCRIPTION
**Pull request by crepererum**
With all the concurrency introduced in #5668, we should be a bit more careful with our "remove if" handling, esp. if a removal is triggered while a load is running concurrently. This change introduces as `remove_if_and_get` helper that ensures this and the querier over to use it. The parquet file and tombstone caches required a bit of a larger change because there the invalidation and the actual GET were kinda separate. We had this separation for the other caches as well at some point and decided that this easily leads to API misuse, so I took this opportunity to "fix" the parquet file and tombstone cache as well.

### CODE REVIEW

File: querier/src/table/mod.rs [hunk 1]
```diff
@@ -236,11 +244,14 @@ impl QuerierTable {
...
                 self.id(),
+                max_parquet_sequence_number,
                 span_recorder.child_span("cache GET parquet_file")
             ),
-            catalog_cache
-                .tombstone()
-                .get(self.id(), span_recorder.child_span("cache GET tombstone"))
+            catalog_cache.tombstone().get(
+                self.id(),
+                max_tombstone_sequence_number,
```
Comment (alamb) [inline querier/src/table/mod.rs:252]: it took me a while to realize that the cache invalidation was now done here by passing in the `max_parquet_sequence_number` and `max_tombstome_sequence_number`to `get` (whereas before the notion of cache invalidation was in the name of the function).

Perhaps we could add a doc comment like

```rust
// NB: pass max parquet/tombstone sequence numbers to `get` 
// to ensure cache is refreshed if we learned about new files/tombstones
```

**Review by alamb — APPROVED** [review]

File: cache_system/src/backend/policy/remove_if.rs [hunk 2]
```diff
@@ -120,6 +122,65 @@ where
...
+            match status {
+                CacheGetStatus::Hit => {
+                    // key existed and no other process loaded it => save to use
+                    return v;
+                }
+                CacheGetStatus::Miss => {
+                    // key didn't exist and we loaded it => save to use
+                    return v;
+                }
+                CacheGetStatus::MissAlreadyLoading => {
```
[the remark this answers is not in the document]
Comment (crepererum) [reply cache_system/src/backend/policy/remove_if.rs:163]: It's technically: "cache miss, but loading was already in progress when request came in". That's a bit long for a enum name :sweat_smile:

I'm open for suggestions though :)

[the remark this answers is not in the document]
Comment (alamb) [reply cache_system/src/backend/policy/remove_if.rs:163]: `MissReloaded`?

File: querier/src/table/mod.rs (hunk 1 above)
Comment (crepererum) [reply querier/src/table/mod.rs:254]: done

File: cache_system/src/backend/policy/remove_if.rs [hunk 3]
```diff
@@ -120,6 +122,65 @@ where
...
+                    if removed {
+                        // key was outdated but there was some loading in process, this may have overlapped with our check
+                        // so our check might have been incomplete => need to re-check
+                        removed = self.remove_if(&k, &predicate);
+                        if removed {
+                            // removed again, so cannot use our result
+                            continue;
+                        } else {
+                            // didn't remove => save to use
```
[the remark this answers is not in the document]
Comment (crepererum) [reply cache_system/src/backend/policy/remove_if.rs]: done (all three of them)

File: cache_system/src/backend/policy/remove_if.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (crepererum) [reply cache_system/src/backend/policy/remove_if.rs:164]: "miss then loaded" is also true for the normal "miss", you get the value back in both cases so it was loaded in the end.

**Pull request merged**